include/lin/references.hpp

```diff
@@ -1,11 +1,450 @@
+#include "references/vector_mapping_reference.hpp"
+#include "references/vector_stream_reference.hpp"
+
+#include <type_traits>
+
+namespace lin {
+namespace internal {
+
+/** @internal
+ *
+ *  @brief Provides a specific tensor type's corresponding mapping reference
+ *         type.
+ *
+ *  @tparam C %Tensor type whose traits are replicated.
+ *  @tparam D Underlying referenced type.
+ *
+ *  @sa internal::stream_reference
+ *
+ *  @ingroup REFERENCES
+ */
+template <class C, class D, typename = void>
+struct mapping_reference;
+
+template <class C, class D>
+using mapping_reference_t = typename mapping_reference<C, D>::type;
+
+template <class C, class D>
+struct mapping_reference<C, D, std::enable_if_t<conjunction<
+    is_matrix<C>, have_same_elements<C, D>>::value>> {
+  typedef MatrixMappingReference<D, C::Traits::rows, C::Traits::cols, C::Traits::max_rows, C::Traits::max_cols> type;
+};
+
+template <class C, class D>
+struct mapping_reference<C, D, std::enable_if_t<conjunction<
+    is_col_vector<C>, have_same_elements<C, D>>::value>> {
+  typedef VectorMappingReference<D, C::VectorTraits::length, C::VectorTraits::max_length> type;
+};
+
+template <class C, class D>
+struct mapping_reference<C, D, std::enable_if_t<conjunction<
+    is_row_vector<C>, have_same_elements<C, D>>::value>> {
+  typedef RowVectorMappingReference<D, C::VectorTraits::length, C::VectorTraits::max_length> type;
+};
+
+/** @internal
+ *
+ *  @brief Provides a specific tensor type's corresponding stream reference
+ *         type.
+ *
+ *  @tparam C %Tensor type whose traits are replicated.
+ *  @tparam D Underlying referenced type.
+ *
+ *  @sa internal::mapping_reference
+ *
+ *  @ingroup REFERENCES
+ */
+template <class C, class D, typename = void>
+struct stream_reference;
+
+template <class C, class D>
+using stream_reference_t = typename stream_reference<C, D>::type;
+
+template <class C, class D>
+struct stream_reference<C, D, std::enable_if_t<conjunction<
+    is_matrix<C>, have_same_elements<C, D>>::value>> {
+  typedef MatrixStreamReference<D, C::Traits::rows, C::Traits::cols, C::Traits::max_rows, C::Traits::max_cols> type;
+};
+
+template <class C, class D>
+struct stream_reference<C, D, std::enable_if_t<conjunction<
+    is_col_vector<C>, have_same_elements<C, D>>::value>> {
+  typedef VectorStreamReference<D, C::VectorTraits::length, C::VectorTraits::max_length> type;
+};
+
+template <class C, class D>
+struct stream_reference<C, D, std::enable_if_t<conjunction<
+    is_row_vector<C>, have_same_elements<C, D>>::value>> {
+  typedef RowVectorStreamReference<D, C::VectorTraits::length, C::VectorTraits::max_length> type;
+};
+}  // namespace internal
+
+/** @brief Creates a mapping reference with default dimensions.
+ *
+ *  @tparam C %Tensor type whose traits are replicated.
+ *  @tparam D Underlying referenced type.
+ *
+ *  @param mapping Underlying mapping.
+ *  @param i       Anchor point row index.
+ *  @param j       Anchor point column index.
+ *
```

Comment:
Anchor point is top left? and if so could you clarify?

Reply:
Good catch - adding this to all the top level functions and internal type constructors.